src/coreclr/jit/importer.cpp

```diff
@@ -3647,6 +3647,110 @@ const char* Compiler::impGetIntrinsicName(CorInfoIntrinsics intrinsicID)
+    assert(sig->numArgs == 1);
+    assert(sig->sigInst.methInstCount == 1);
+
+    GenTree* fieldTokenNode = impStackTop(0).val;
+
+    //
+    // Verify that the field token is known and valid.  Note that it's also
+    // possible for the token to come from reflection, in which case we cannot do
+    // the optimization and must therefore revert to calling the helper.  You can
+    // see an example of this in bvt\DynIL\initarray2.exe (in Main).
+    //
+
+    // Check to see if the ldtoken helper call is what we see here.
+    if (fieldTokenNode->gtOper != GT_CALL || (fieldTokenNode->AsCall()->gtCallType != CT_HELPER) ||
+        (fieldTokenNode->AsCall()->gtCallMethHnd != eeFindHelper(CORINFO_HELP_FIELDDESC_TO_STUBRUNTIMEFIELD)))
+    {
+        return nullptr;
+    }
+
+    // Strip helper call away
+    fieldTokenNode = fieldTokenNode->AsCall()->gtCallArgs->GetNode();
+    if (fieldTokenNode->gtOper == GT_IND)
+    {
+        fieldTokenNode = fieldTokenNode->AsOp()->gtOp1;
+    }
+
+    // Check for constant
+    if (fieldTokenNode->gtOper != GT_CNS_INT)
+    {
+        return nullptr;
+    }
+
+    CORINFO_FIELD_HANDLE fieldToken = (CORINFO_FIELD_HANDLE)fieldTokenNode->AsIntCon()->gtCompileTimeHandle;
+    if (!fieldTokenNode->IsIconHandle(GTF_ICON_FIELD_HDL) || (fieldToken == nullptr))
+    {
+        return nullptr;
+    }
+
+    CORINFO_CLASS_HANDLE fieldOwnerHnd = info.compCompHnd->getFieldClass(fieldToken);
+
+    CORINFO_CLASS_HANDLE fieldClsHnd;
+    var_types fieldElementType = JITtype2varType(info.compCompHnd->getFieldType(fieldToken, &fieldClsHnd, fieldOwnerHnd));
+    unsigned totalFieldSize;
+
+    // Most static initialization data fields are of some structure, but it is possible for them to be of various primitive types as well
+    if (fieldElementType == var_types::TYP_STRUCT)
+    {
+        totalFieldSize = info.compCompHnd->getClassSize(fieldClsHnd);
+    }
+    else
+    {
+        totalFieldSize = genTypeSize(fieldElementType);
+    }
+
+    // Limit to primitive or enum type - see ArrayNative::GetSpanDataFrom()
+    CORINFO_CLASS_HANDLE targetElemHnd  = sig->sigInst.methInst[0];
+    if (info.compCompHnd->getTypeForPrimitiveValueClass(targetElemHnd) == CORINFO_TYPE_UNDEF)
+    {
+        return nullptr;
+    }
+
+    const unsigned targetElemSize = info.compCompHnd->getClassSize(targetElemHnd);
+    assert(targetElemSize != 0);
+
+    const unsigned count = totalFieldSize / targetElemSize;
+    if (count == 0)
+    {
+        return nullptr;
+    }
+
+    void* data = info.compCompHnd->getArrayInitializationData(fieldToken, totalFieldSize);
+    if (!data)
+    {
+        return nullptr;
+    }
+
+    //
+    // Ready to commit to the work
+    //
+
+    impPopStack();
+
+    // Turn count and pointer value into constants.
+    GenTree* spanElemCount = gtNewIconNode(count, TYP_INT);
+    GenTree* spanPointerField = gtNewIconHandleNode((size_t)data, GTF_ICON_CONST_PTR);
+
+    // Construct ReadOnlySpan<T> to return.
```

Comment:
I'd personally emit just a call to `System.ReadOnlySpan'1[MyType][MyType]..ctor` and let the inliner do the work. Something like:
```
*  CALL      void   System.ReadOnlySpan`1[MyType][MyType]..ctor
+--*  ADDR      byref 
|  \--*  LCL_VAR   struct<System.ReadOnlySpan`1[MyType], %size%> spanThisTmp  
+--*  CNS_INT(h) long   %spanPointerField%
\--*  CNS_INT   int    %spanElemCount%
```
but maybe inlining it by hands produces more optimal and TP-wise IR, let's see.

Reply:
For now this code is suboptimal. e.g.:
```csharp
public int M(int index)
{
    ReadOnlySpan<int> data = new int[] { 1, 2, 3, 4 };
    return data[index];
}
```
your code emits:
```asm
G_M30088_IG01:              ;; offset=0000H
       4883EC38             sub      rsp, 56
       33C0                 xor      eax, eax
       4889442428           mov      qword ptr [rsp+28H], rax
       4889442430           mov      qword ptr [rsp+30H], rax
                                                ;; bbWeight=1    PerfScore 2.50
G_M30088_IG02:              ;; offset=0010H
       C744243004000000     mov      dword ptr [rsp+30H], 4
       48B850201D6C1C020000 mov      rax, 0x21C6C1D2050
       4889442428           mov      bword ptr [rsp+28H], rax
       488B442428           mov      rax, bword ptr [rsp+28H]
       8B542430             mov      edx, dword ptr [rsp+30H]
       3BCA                 cmp      ecx, edx
       730A                 jae      SHORT G_M30088_IG04
       8BD1                 mov      edx, ecx
       8B0490               mov      eax, dword ptr [rax+4*rdx]
                                                ;; bbWeight=1    PerfScore 7.75
G_M30088_IG03:              ;; offset=0039H
       4883C438             add      rsp, 56
       C3                   ret
                                                ;; bbWeight=1    PerfScore 1.25
G_M30088_IG04:              ;; offset=003EH
       E84D8A665F           call     CORINFO_HELP_RNGCHKFAIL
       CC                   int3
                                                ;; bbWeight=0    PerfScore 0.00
```
e.g. `spanPointerField` spilled to stack unnecessary. Maybe it's GT_COMMA's fault and `impAppendTree` should be used to produce statements instead.

Reply:
@SingleAccretion is right, raw IND operations make IR not struct-promotion friendly.

Reply:
Resolved in https://github.com/dotnet/runtime/pull/61079#discussion_r741159096